src/sinks/splunk_hec/logs/sink.rs

```diff
@@ -150,17 +153,31 @@ pub fn process_log(
 
     let host = log.get(host_key).cloned();
 
-    let timestamp = match log.remove(log_schema().timestamp_key()) {
-        Some(Value::Timestamp(ts)) => ts,
-        _ => chrono::Utc::now(),
+    let timestamp = if timestamp_key.is_empty() {
+        None
+    } else {
+        match log.remove(timestamp_key) {
+            Some(Value::Timestamp(ts)) => {
+                // set nanos in log if valid timestamp in event and timestamp_nanos_key is configured
+                if let Some(key) = timestamp_nanos_key {
+                    log.try_insert(path!(key), ts.timestamp_subsec_nanos() % 1_000_000);
+                }
+                Some((ts.timestamp_millis() as f64) / 1000f64)
+            }
+            Some(value) => {
+                warn!(
+                    "timestamp_key is not valid timestamp: type is {}. Ignoring and allowing splunk to set timestamp.",
+                    value.kind_str()
+                );
```

Comment:
👍 to these warnings. We should add internal events for them though. You can see the currently defined events for the `splunk_hec_logs` sink here: https://github.com/vectordotdev/vector/blob/master/src/internal_events/splunk_hec.rs . These are used with the `emit!()` macro you may have seen.

Reply:
Nice. Took a stab at that. I see a lot of references to calling `counter!()` in the intermal events implementation. I left that out for now...

```
    pub struct SplunkEventTimestampWarning<'a> {
        message: &'a str,
    }

    impl<'a> InternalEvent for SplunkEventTimestampWarning<'a> {
        fn emit(self) {
            warn!(message = %self.message);
        }
    }
```

Reply:
This is what it looks like now, with my changes to make it an internal event.

```
running 1 test
2022-06-02T22:25:32.172704Z  WARN vector::internal_events::splunk_hec::sink: timestamp_key is not valid timestamp: type is integer. Ignoring and allowing splunk to set timestamp.
2022-06-02T22:25:32.173662Z  WARN vector::internal_events::splunk_hec::sink: No timestamp was not found. Allowing splunk to set timestamp.
test sinks::splunk_hec::logs::tests::splunk_encode_log_event_json_timestamps ... ok
```